src/Microsoft.AspNet.Server.Kestrel/Infrastructure/ThreadPoolActions.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Server.Kestrel.Http;
+
+namespace Microsoft.AspNet.Server.Kestrel.Infrastructure
+{
+    public class ThreadPoolActions
+    {
+        private readonly IKestrelTrace _log;
+
+        private readonly WaitCallback _runAction;
+        private readonly WaitCallback _completeTcs;
+        private readonly WaitCallback _abortFrame;
+        private readonly WaitCallback _returnBlocks;
+        private readonly WaitCallback _returnWrittenBlocks;
+
+        public ThreadPoolActions(IKestrelTrace log)
+        {
+            _log = log;
+
+            // Curry and capture log in closures once
+            _runAction = (o) =>
+            {
+                try
+                {
+                    ((Action)o)();
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _completeTcs = (o) =>
+            {
+                try
+                {
+                    ((TaskCompletionSource<object>)o).SetResult(null);
```

Comment:
It stops the setting of the value throwing rather than the synchronous running of the continuations from throwing. From digging into Task & tcs its probably worthwhile anyway as `SetResult` calls `TrySetResult`; then decides to throw based on the return.

However this would be the top of the stack the the continuations (which is why I think its put on the threadpool); so I think the try catch is still needed.